### DISCUSSION TITLE
How to manage multiple models correctly?

### DISCUSSION THREAD

**Discussion by Nuthin2do** [CodinGame/monaco-vscode-api#508] [Q&A] [answered]
Hi again,

i would like to know what the recommanded way of managing multiple models  is for this API. Previously with the plain monaco-editor I created a model and stored in within a Map and when I wanted to change the active model then i used a modelId, manually created, searched for it in the Map and retrieved the values for the model and the model itself from there.

The Map i used to store the models with their values and the modelID as the key:

`public modelStates: Map<string, ModelState> = new Map();`

The ModelState type:

```
export class ModelState implements iModelState {
  modelId: string;
  initialCode: string;
  code: string;
  language: string;
  dtsFiles: DtsFile[];
  currentLineNumber: number = -1;
  lineState: ILineState;
  scrollPosition: number | null;
  model: monaco.editor.ITextModel;
  isDirty: boolean = false;

  constructor(
    modelId: string,
    initialCode: string,
    code: string,
    language: string,
    dtsFiles: DtsFile[],
    lineState: ILineState,
    scrollPosition: number | null
  ) {
    this.modelId = modelId;
    this.initialCode = initialCode;
    this.code = code;
    this.language = language;
    this.dtsFiles = dtsFiles;
    this.lineState = lineState;
    this.scrollPosition = scrollPosition ?? null;
    this.model = monaco.editor.createModel(code || "", language);
  }
}
```

and the methods i used to create and update the map and to set the activeModel:

```
public add(modelState: ModelState): string {
    const modelId =
      modelState.modelId ?? "mid" + Math.random().toString(8).slice(2);
    const model = monaco.editor.createModel(
      modelState.code,
      modelState.language
    );

    modelState.initialCode = modelState.code;
    modelState.modelId = modelId;
    modelState.model = model;

    if (modelState.lineState === null && this.editor !== null) {
      modelState.lineState = new LineState();
    }

    this.modelStates.set(modelId, modelState);

    return modelId;
  }

  public setActiveModel(modelId: string) {
    const modelState = this.modelStates.get(modelId);
    if (this.editor && modelState) {
      this.editor.setModel(modelState.model);
      this.activeModelID = modelId;
    }

    if (modelState?.lineState === null && this.editor !== null) {
      modelState.lineState = new LineState();
    }

    if (modelState !== undefined && modelState?.lineState !== null) {
      this.setActiveLineState(modelState.lineState);
    }
  }
```

For the scenario i don't have the VSCode environment like in the demo. I just have the editor with its model (and some services of course). The change of models is triggered outside of the editor in C#. Previously i did create a ModelState in C# and send it to typescript and used the methods i described up there.

Now:

When i create a model as described here (https://github.com/CodinGame/monaco-vscode-api?tab=readme-ov-file#model-creation) do i have to store the modelRef with the fileUri manually (like i did before in a Map) to be able to get the right model back when i change the active model? Assuming i get the modelRef i want based on the modelID, do i just have to call setModel from the IStandaloneEditor?

Can you provide me a small example or some directions?

Thank you very much.

**Comment by CGNonofr** [comment, marked as answer]
Storing the model reference looks like a good replacement idea

but you don't really need it, you can call createModelReference at any time, as soon as the file is registered in the virtual filesystem, and release it as soon as you don't need it directly

Note that as soon as there at least one reference to a model, the model is considered "active" (so for instance, the LSP server will receive a didOpen on it)

**Reply by Nuthin2do** [reply]
Thanks for your reply.

> `but you don't really need it, you can call createModelReference at any time, as soon as the file is registered in the virtual filesystem, and release it as soon as you don't need it directly`

So, that means all i need to store is the fileUri and based on it i can get the right model with its state like code, the initial code, language and sp as soon as i call createModelReference?

**Reply by CGNonofr** [reply]
You understood correctly !
